src/raftstore/store/worker/region.rs

```diff
@@ -248,31 +258,160 @@ impl SnapContext {
         }
 
         timer.observe_duration();
+        region
     }
 
     fn handle_destroy(&mut self, region_id: u64, start_key: Vec<u8>, end_key: Vec<u8>) {
-        info!(
-            "[region {}] deleting data in [{}, {})",
-            region_id,
-            escape(&start_key),
-            escape(&end_key)
-        );
+        if let Err(e) = util::delete_all_files_in_range(&self.kv_db, &start_key, &end_key) {
+            error!(
+                "[region {}] failed to delete files in [{}, {}): {:?}",
+                region_id,
+                escape(&start_key),
+                escape(&end_key),
+                e
+            );
+            return;
+        }
         if let Err(e) =
             util::delete_all_in_range(&self.kv_db, &start_key, &end_key, self.use_delete_range)
         {
             error!(
-                "failed to delete data in [{}, {}): {:?}",
+                "[region {}] failed to delete data in [{}, {}): {:?}",
+                region_id,
                 escape(&start_key),
                 escape(&end_key),
                 e
             );
+        } else {
+            info!(
+                "[region {}] succeed in deleting data in [{}, {})",
+                region_id,
+                escape(&start_key),
+                escape(&end_key),
+            );
+        }
+    }
+}
+
+#[derive(Default)]
+struct PendingDeleteRanges {
+    ranges: BTreeMap<Vec<u8>, (Vec<u8>, u64, time::Instant)>, // start_key -> (end_key, region_id, ttl)
+}
+
+impl PendingDeleteRanges {
+    // find out the ranges that need to be trimed or discarded
+    fn find_overlapped_ranges(&self, start_key: &[u8], end_key: &[u8]) -> Vec<(Vec<u8>, Vec<u8>)> {
+        let mut ranges = Vec::new();
+        // find the first range that may overlap with [start_key, end_key)
+        let sub_range = self.ranges.range((Unbounded, Excluded(start_key.to_vec())));
+        if let Some((s_key, &(ref e_key, _, _))) = sub_range.last() {
+            if e_key > &start_key.to_vec() {
+                ranges.push((s_key.clone(), e_key.clone()));
+            }
+        }
+
+        // find the rest ranges that overlap with [start_key, end_key)
+        for (s_key, &(ref e_key, _, _)) in self.ranges
+            .range((Included(start_key.to_vec()), Excluded(end_key.to_vec())))
+        {
+            ranges.push((s_key.clone(), e_key.clone()));
+        }
+        ranges
+    }
+
+    // filter out the overlap in ranges, after this function the ranges will not overlap
+    pub fn remove(&mut self, start_key: &[u8], end_key: &[u8]) {
```

Comment:
IMO, this is complicated and error-prone. If overlap will not happen frequently, I think we can just remove any overlapped ranges, instead of updating these ranges.
/cc @zhangjinpeng1987

Reply:
So this means that we need to clean some of the overlapped ranges when tikv restarts?

Reply:
I discussed with @zhangjinpeng1987 and this may lead to a lot of stale peers.

Reply:
I guess @huachaohuang means still removing the overlapped region, but not using following complex way.

Reply:
I thinks he means remove it all, but somestimes the region should not be removed all, only part of the region should be removed, otherwise some part of the stale peer won't be cleaned until tikv restarts.

Reply:
part of the region, em,  Can any case cause this?

Reply:
I checked the log, sometimes it will occur, even it doesn't, we should cover it anyway.

Reply:
why not removing the whole overlapped region at first?